content/draft/cpython-tail-call.md

```diff
@@ -0,0 +1,263 @@
+
+```c++
+static void *opcode_table[256] = {
+    [OP_IMM] = &&TARGET_IMM,
+    [OP_ADD] = &&TARGET_ADD,
+    // etc
+};
+
+#define DISPATCH() goto *opcode_table[bytecode[pc++]]
+
+DISPATCH();
+
+TARGET_IMM: {
+    // push an immediate onto the stack
+    DISPATCH();
+}
+TARGET_ADD: {
+    // handle the add
+    DISPATCH();
+}
+```
+
+## The bug
+
+For performance reasons (performance of the compiler, not the generated code), it turns out that clang and LLVM, internally, actually turn the latter code code, internally, into something that looks structurally more like the original `while(){switch}`-based implementation! That is, they merge all of the "computed goto" statements into a **single** [indirectbr][indirectbr] instruction in the LLVM IR, and branch to that instruction from the end of each opcode definition.
+
+Then, during code generation, they implement a pass which implement "tail duplication" to copy the `indirectbr` back into every basic block, in order to restore the original intent. You can read more [in this old blog post from the LLVM blog][llvm-computed-goto].
+
+Clang 19 included [a tweak to that tail-duplication pass][llvm-bugpr], which added some limits in order to try not to blow up the size of the internal IR.
+
+Unfortunately, those limits resulted in `clang` **leaving all of the dispatch jumps merged**, and undoing the entire benefit of the computed-goto-based implementation! This bug was [first identified][llvm-bug] by another language implementation with a similar interpreter loop, but had not been known (as far as I can find) to affect CPython.
+
+In addition to the performance impact, we can observe the bug directly by disassembling the resulting object code and counting the number of distinct indirect jumps:
+
+```shell
+$ objdump -S --disassemble=_PyEval_EvalFrameDefault ${clang18}/bin/python3.14 | \
+  egrep -c 'jmp\s+\*'
+332
+
+$ objdump -S --disassemble=_PyEval_EvalFrameDefault ${clang19}/bin/python3.14 | \
+  egrep -c 'jmp\s+\*'
+3
+```
+
+### Further weirdness
+
+In fact, there's something even odder going on, which I can't fully explain.
+
+Classically, the optimization of replicating the bytecode dispatch into each opcode is cited as speeding up interpreters anywhere from [20%](https://github.com/python/cpython/blob/c718c6be0f82af5eb0e57615ce323242155ff014/Misc/HISTORY#L15252-L15255) to [100%](https://link.springer.com/content/pdf/10.1007/3-540-44681-8_59.pdf). However, on modern processors with improved branch predictors, [more recent work](https://inria.hal.science/hal-01100647/document) finds a much smaller speedup, on the order of 2-4%.
+
+Thus, I can't explain why the clang19 performance regression is as steep as it is. I can confirm that it's fixed by re-enabling tail duplication, but I cannot explain the 10% slowdown **only* as a result of tail merging.
```

Comment:
Yeah same here :(. Maybe the current indirectbr stuff _worsens_ the branch predictor rather than just being a neutral party?